src/Components/Components/src/Forms/EditContext.cs

```diff
@@ -0,0 +1,185 @@
+
+        /// <summary>
+        /// Gets the model object for this <see cref="EditContext"/>.
+        /// </summary>
+        public object Model { get; }
+
+        /// <summary>
+        /// Signals that the value for the specified field has changed.
+        /// </summary>
+        /// <param name="fieldIdentifier">Identifies the field whose value has been changed.</param>
+        public void NotifyFieldChanged(FieldIdentifier fieldIdentifier)
+        {
+            GetFieldState(fieldIdentifier, ensureExists: true).IsModified = true;
+            OnFieldChanged?.Invoke(this, fieldIdentifier);
+        }
+
+        /// <summary>
+        /// Signals that some aspect of validation state has changed.
+        /// </summary>
+        public void NotifyValidationStateChanged()
+        {
+            OnValidationStateChanged?.Invoke(this, null);
+        }
+
+        /// <summary>
+        /// Clears any modification flag that may be tracked for the specified field.
+        /// </summary>
+        /// <param name="fieldIdentifier">Identifies the field whose modification flag (if any) should be cleared.</param>
+        public void MarkAsUnmodified(FieldIdentifier fieldIdentifier)
+        {
+            if (_fieldStates.TryGetValue(fieldIdentifier, out var state))
+            {
+                state.IsModified = false;
+            }
+        }
+
+        /// <summary>
+        /// Clears all modification flags within this <see cref="EditContext"/>.
+        /// </summary>
+        public void MarkAsUnmodified()
+        {
+            foreach (var state in _fieldStates.Values)
+            {
+                state.IsModified = false;
+            }
+        }
+
+        /// <summary>
+        /// Determines whether any of the fields in this <see cref="EditContext"/> have been modified.
+        /// </summary>
+        /// <returns>True if any of the fields in this <see cref="EditContext"/> have been modified; otherwise false.</returns>
+        public bool IsModified()
+            // If necessary, we could consider caching the overall "is modified" state and only recomputing
+            // when there's a call to NotifyFieldModified/NotifyFieldUnmodified
+            => _fieldStates.Values.Any(state => state.IsModified);
+
+        /// <summary>
+        /// Gets the current validation messages across all fields.
+        ///
+        /// This method does not perform validation itself. It only returns messages determined by previous validation actions.
+        /// </summary>
+        /// <returns>The current validation messages.</returns>
+        public IEnumerable<string> GetValidationMessages()
+            // Since we're only enumerating the fields for which we have a non-null state, the cost of this grows
+            // based on how many fields have been modified or have associated validation messages
+            => _fieldStates.Values.SelectMany(state => state.GetValidationMessages());
```

Comment:
Note: all of these cases of linq allocate - a `foreach` doesn't (for `Dictionary<>.ValueCollection`)

Reply:
I'm not laser-focused on the perf of stuff yet, just trying to build a mental model about what tradeoffs are possible.

Reply:
Good point. I've switched this to `yield return` now, so as to retain the ability to short-circuit from the caller's side.